src/Language/AST/ArgumentNode.php

```diff
@@ -1,17 +1,23 @@
 <?php
+
+declare(strict_types=1);
+
 namespace GraphQL\Language\AST;
 
 class ArgumentNode extends Node
 {
+    /** @var string */
     public $kind = NodeKind::ARGUMENT;
 
-    /**
-     * @var ValueNode
-     */
+    /** @var ValueNode */
     public $value;
 
-    /**
-     * @var NameNode
-     */
+    /** @var NameNode */
     public $name;
+
+
+    public function getKind() : string
+    {
+        return NodeKind::ARGUMENT;
+    }
```

Comment:
We can leave them available publicly but create getters compliant with interfaces we define. `getKind()` is the first getter I defined.

I planned to continue with defining other properties via interfaces so we can better leverage type-safety.

![image](https://user-images.githubusercontent.com/327717/44336814-e92ef680-a478-11e8-8593-e14ffd0c98f5.png)

![image](https://user-images.githubusercontent.com/327717/44336823-f0ee9b00-a478-11e8-9d8b-761252ce2c45.png)


So I suggest to rework it as
```php
    public function getKind() : string
    {
        return $this->kind;
    }
```

Performance shouldn't suffer as long as the property stays public https://steemit.com/php/@crell/php-use-associative-arrays-basically-never

Reply:
BTW the reason I'm doing all this is to (in the end) add types to this lib and make it better prepared for future php releases so it can leverage all the new features that are going to come (with performance gains as well :)

Reply:
Actually, your link says that getters do add a quite significant overhead to performance (8.21 vs 11.08 seconds in tests).

And that's actually the main reason why AST is designed like this (without getters). PHP is not the best language for projects like this, so we really have to sacrifice some of the best practices for performance gains.

So I would actually suggest reverting the getKind() getter. And it's better to separate the work on the code style from changes to interfaces and logic, because those require some benchmarking first.

Reply:
In the link it is said that 11.08s avg is for private properties but I suggested to keep having public properties and adding getters for gaining API advantage via interfaces.

Also, I understand that making such changes is better for separate PRs. I'll try to clean it up from this one and open PR when we can reopen this discussing, alright?

Reply:
Sure